vllm/lora/layers.py

```diff
@@ -952,17 +1027,29 @@ def from_layer(
         layer: nn.Module,
         max_loras: int,
         lora_config: LoRAConfig,
+        packed_modules_list: List,
         model_config: Optional[PretrainedConfig] = None) -> BaseLayerWithLoRA:
     supported_layer_types = {
-        VocabParallelEmbedding: VocabParallelEmbeddingWithLoRA,
-        ColumnParallelLinear: ColumnParallelLinearWithLoRA,
-        QKVParallelLinear: QKVParallelLinearWithLora,
-        MergedColumnParallelLinear: MergedColumnParallelLinearWithLoRA,
-        RowParallelLinear: RowParallelLinearWithLoRA,
+        VocabParallelEmbedding:
+        VocabParallelEmbeddingWithLoRA,
+        ColumnParallelLinear:
+        ColumnParallelLinearWithLoRA,
+        QKVParallelLinear:
+        (QKVParallelLinearWithLora, RefQKVParallelLinearWithLora),
+        MergedColumnParallelLinear: (MergedColumnParallelLinearWithLoRA,
+                                     RefMergedColumnParallelLinearWithLoRA),
+        RowParallelLinear:
+        RowParallelLinearWithLoRA
     }
     for src_layer_type, lora_layer_type in supported_layer_types.items():
         if type(layer) is src_layer_type:  # pylint: disable=unidiomatic-typecheck
-            ret = lora_layer_type(layer)
+            if isinstance(lora_layer_type, tuple):
+                if len(packed_modules_list) > 1:
+                    ret = lora_layer_type[0](layer)
+                else:
+                    ret = lora_layer_type[1](layer)
+            else:
+                ret = lora_layer_type(layer)
```

Comment:
this is too brittle. how about we add a class method to the lora layers that would act as a condition to be called here? eg. `can_replace_src_layer`